process/process.go

```diff
@@ -0,0 +1,380 @@
+	}
+	p.StartRound(0)
+	return p
+}
+
+// MarshalJSON implements the `json.Marshaler` interface for the Process type,
+// by marshaling its isolated State.
+func (p Process) MarshalJSON() ([]byte, error) {
+	return json.Marshal(p.state)
+}
+
+// UnmarshalJSON implements the `json.Unmarshaler` interface for the Process
+// type, by unmarshaling its isolated State.
+func (p *Process) UnmarshalJSON(data []byte) error {
+	return json.Unmarshal(data, &p.state)
+}
+
+// StartRound is safe for concurrent use. See
+// https://arxiv.org/pdf/1807.04938.pdf for more information.
+func (p *Process) StartRound(round block.Round) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	p.startRound(round)
+}
+
+// HandleMessage is safe for concurrent use. See
+// https://arxiv.org/pdf/1807.04938.pdf for more information.
+func (p *Process) HandleMessage(m Message) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	switch m := m.(type) {
+	case *Propose:
+		p.handlePropose(m)
+	case *Prevote:
+		p.handlePrevote(m)
+	case *Precommit:
+		p.handlePrecommit(m)
+	}
+}
+
+func (p *Process) startRound(round block.Round) {
+	p.state.CurrentRound = round
+	p.state.CurrentStep = StepPropose
+	if p.signatory.Equal(p.scheduler.Schedule(p.state.CurrentHeight, p.state.CurrentRound)) {
+		var proposal block.Block
+		if p.state.ValidBlock.Hash() != block.InvalidHash {
+			proposal = p.state.ValidBlock
+		} else {
+			proposal = p.proposer.BlockProposal(p.state.CurrentHeight, p.state.CurrentRound)
+		}
+		p.broadcaster.Broadcast(NewPropose(
+			p.state.CurrentHeight,
+			p.state.CurrentRound,
+			proposal,
+			p.state.ValidRound,
+		))
+	} else {
+		p.scheduleTimeoutPropose(p.state.CurrentHeight, p.state.CurrentRound, p.timer.Timeout(StepPropose, p.state.CurrentRound))
+	}
+}
+
+func (p *Process) handlePropose(propose *Propose) {
+	_, firstTime, _, _ := p.state.Proposals.Insert(propose)
+
+	// upon Propose{currentHeight, currentRound, block, -1}
+	if propose.Height() == p.state.CurrentHeight && propose.Round() == p.state.CurrentRound && propose.ValidRound() == block.InvalidRound {
+		// from Schedule{currentHeight, currentRound}
+		if propose.Signatory().Equal(p.scheduler.Schedule(p.state.CurrentHeight, p.state.CurrentRound)) {
+			// while step = StepPropose
+			if p.state.CurrentStep == StepPropose {
+				if p.validator.IsBlockValid(propose.Block()) && (p.state.LockedRound == block.InvalidRound || p.state.LockedBlock.Equal(propose.Block())) {
+					p.broadcaster.Broadcast(NewPrevote(
+						p.state.CurrentHeight,
+						p.state.CurrentRound,
+						propose.Block().Hash(),
+					))
+				} else {
+					p.broadcaster.Broadcast(NewPrevote(
+						p.state.CurrentHeight,
+						p.state.CurrentRound,
+						block.InvalidHash,
+					))
+				}
+				p.state.CurrentStep = StepPrevote
+			}
+		}
+	}
+
+	p.checkProposeInCurrentHeightAndRoundWithPrevotes()
+	if firstTime {
+		p.checkProposeInCurrentHeightAndRoundWithPrevotesForTheFirstTime()
+	}
+	p.checkProposeInCurrentHeightWithPrecommits(propose.Round())
```

Comment:
This is already accounted for by Tendermint. Missing blocks won't cause safety/liveliness issues (liveliness will be affected when 1/3+ are missing blocks). But, this is also something that has to be handled at the networking level with broadcasting retries.

Reply:
Yes, correct me if I am wrong. 
Tendermint assumes that all messages will eventually be gossiped to all other processes within a certain time.
```Gossip communication:If a correct process 'p' receives some message 'm' at time 't', all correct processes will receive 'm' before max{t, GST}+ ∆.``` 

In our case, what if a replica goes offline for half an hour? This would mean that it will have to sync up its blockchain before participating in consensus?

Reply:
@divyakoshy yes, that's correct. But, that is only necessary if more than `f` replicas exhibit that behaviour at the same time (because we can just think about them as incorrect processes). Obviously, though, it is ideal that we do a full sync when replicas reboot. This is outside the concerns of Hyperdrive though (which leaves all networking to the programmer implementing the interfaces).